Fix bind usage in example in addEventListener doc

akanshgulati opened the pull request:
Remove bind directly used in addEventListener method to avoid memory leak as removeListener on current method will not work

<!-- Please provide the following information to help us review this PR: -->

> What was wrong/why is this fix needed? (quick summary only)

> MDN URL of main page changed

> Issue number (if there is an associated issue)

> Anything else that could help us review it

akanshgulati committed:
Move bind to different line for onclick1 method

Remove bind directly used in addEventListener method to avoid memory leak as removeListener on current method will not work

Ryuno-Ki commented on files/en-us/web/api/eventtarget/addeventlistener/index.html:
```diff
@@ -593,11 +593,14 @@ <h4 id="Specifying_this_using_bind">Specifying "this" using bind()</h4>
   this.onclick1 = function(event) {
     console.log(this.name); // undefined, as |this| is the element
   };
+  // bind creates a new instance of the method and we can assign it to own method if `this` instance is fixed for the method
```
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/bind says

> The bind() method creates a new function that, when called, has its this keyword set to the provided value, with a given sequence of arguments preceding any provided when the new function is called.

A new instance would be `new MyClass()`. A method is often associated with a class.
So I'm a bit unhappy with the wording here, but can get the sentiment.

Ryuno-Ki reviewed (commented):
I have some thoughts on this.

Ryuno-Ki commented:
@sideshowbarker Friendly ping

sideshowbarker reviewed (commented):
I agree the existing example in the MDN article is not great — and after looking at it I feel like I’d kind of like to rewrite the whole thing — but the change in this PR as currently written doesn’t seem like it will result in greater clarity for readers about that example.

The text added also not doesn’t make clear what problem this change is solving; the issue description says, *Remove bind directly used in addEventListener method to avoid memory leak as removeListener on current method will not work* — but reading through the example, it’s not clear that’s what it’s doing. Maybe it needs some additional code that shows removeListener being called somewhere else outside that function, or something.

Otherwise, even though that existing example is maybe not so great, there’s not actually an bug in it. It will work. But the problem with it is what happens if/when somebody using that code wants to remove that event listener somewhere outside that function.

akanshgulati commented on files/en-us/web/api/eventtarget/addeventlistener/index.html:
```diff
@@ -593,11 +593,14 @@ <h4 id="Specifying_this_using_bind">Specifying "this" using bind()</h4>
   this.onclick1 = function(event) {
     console.log(this.name); // undefined, as |this| is the element
   };
+  // bind creates a new instance of the method and we can assign it to own method if `this` instance is fixed for the method
```
@sideshowbarker I agree with the words not appropriate, I was more focused on stating the problem first. We can think of better comment and notification for the user to understand the reason.

akanshgulati committed:
Update index.html to move bind method

Move bind method from addEventListener callback to separate line to have the reference for removing it

akanshgulati commented:
@Ryuno-Ki Please check it again, I have tried to add more sense to that comment and description.

Ryuno-Ki commented on files/en-us/web/api/eventtarget/addeventlistener/index.html:
```diff
@@ -579,9 +579,8 @@ <h3 id="The_value_of_this_within_the_handler">The value of "this" within the han
-<p>The {{jsxref("Function.prototype.bind()")}} method lets you specify the value that
-  should be used as <code>this</code> for all calls to a given function. This lets you
-  easily bypass problems where it's unclear what <code>this</code> will be, depending on
+<p>The {{jsxref("Function.prototype.bind()")}} method creates a new function for a given function that lets you specify the value that
```
I get what you mean, but I think, it's not correct.

I am reading „creates a new function” as a return value of `bind`.
Instead, the function it's called on is mutated, i.e. I don't have to assign a return value.

Ryuno-Ki reviewed (commented):
The example looks good now!
But we'll need to get the explanation right.

sideshowbarker commented on files/en-us/web/api/eventtarget/addeventlistener/index.html:
```diff
@@ -579,9 +579,8 @@ <h3 id="The_value_of_this_within_the_handler">The value of "this" within the han
-<p>The {{jsxref("Function.prototype.bind()")}} method lets you specify the value that
-  should be used as <code>this</code> for all calls to a given function. This lets you
-  easily bypass problems where it's unclear what <code>this</code> will be, depending on
+<p>The {{jsxref("Function.prototype.bind()")}} method creates a new function for a given function that lets you specify the value that
+  should be used as <code>this</code> for all calls. This easily bypass problems where it's unclear what <code>this</code> will be, depending on
```
```suggestion
<p>The {{jsxref("Function.prototype.bind()")}} method lets you establish a fixed 
  <code>this</code> context for all subsequent calls — bypassing problems where it's unclear what <code>this</code> will be, depending on
```

sideshowbarker commented on files/en-us/web/api/eventtarget/addeventlistener/index.html:
```diff
@@ -593,11 +592,16 @@ <h4 id="Specifying_this_using_bind">Specifying "this" using bind()</h4>
     console.log(this.name); // 'Something Good', as |this| is bound to newly created object
   };
+  
+  // bind creates a new method with fixed `this` context assigned to onclick2
```
```suggestion
  // bind causes a fixed `this` context to be assigned to onclick2
```

sideshowbarker committed:
Update files/en-us/web/api/eventtarget/addeventlistener/index.html

sideshowbarker committed:
Update files/en-us/web/api/eventtarget/addeventlistener/index.html

akanshgulati commented:
@sideshowbarker Thanks for updating it, but the whole point of raising the issue was to avoid using `bind` directly inside the addEventListener callback method in the example.

sideshowbarker commented:
> @sideshowbarker Thanks for updating it, but the whole point of raising the issue was to avoid using `bind` directly inside the addEventListener callback method in the example.

I'm confused by that statement. Are you saying that the code in the patch I merged doesn't avoid using `bind` directly inside the addEventListener callback method in the example?

It's the same code you put into the patch, right? I didn't change the code, did I? My only suggestions and changes were to the comments. I didn't alter the code, and neither did any other reviewer, right?

So what are you saying?

Are you unhappy with the patch that got merged?

Do you want to revert the merge?

If not, what else would you like to have happen at this point?

akanshgulati commented:
@sideshowbarker Thanks for quick reply. Actually I checked the MR status coming as being merged and assumed that it has gone live too on production. My bad. Seems like the changes are merged but not went live yet.

sideshowbarker commented:
> @sideshowbarker Thanks for quick reply. Actually I checked the MR status coming as being merged and assumed that it has gone live too on production. My bad. Seems like the changes are merged but not went live yet.

Ah, ok, cool

Yeah, it can take up to 48 hours before the changes will show up in the production site.
